### PR TITLE
load balancer and its configurations resource for terraform

### PR DESCRIPTION
**Pull request by Pushpalatha555**
Validated the Terraform For CREATE/DELETE/READ for load balancer and its configurations like monitor, profile ,pool, virtual server

LB-CREATE

LB-DELETE

LB-Monitor-CREATE

LB-Monitor-DELETE

LB-POOL-CREATE

LB-POOL-DELETE

LB-PROFILE-CREATE

LB-PROFILE-DELETE

LB-VS-CREATE

LB-VS-DELETE

### CODE REVIEW

**Review by gandharvas — DISMISSED** [review]

**Review by gandharvas — COMMENTED** [review]
I hope the required acceptance tests are to be added to the PR.

**Review by ravikumar29524 — APPROVED** [review]

**Review by ravikumar29524 — APPROVED** [review]

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
+
+resource "hpegl_vmaas_load_balancer" "tf_load_balancer" {
+  name  = "lab-1"       
+  description  = "creating load balancer for test"
+  enabled      =    true        
+  visibility   = "public"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:7]: wont be applicable

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb_pool.tf [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
+
+resource "hpegl_vmaas_load_balancer_pool" "test_lb_pool" {
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb_pool.tf:3]: I guess this is created for a load balancer, we need to take the load balacers details here as the input. https://apidocs.morpheusdata.com/#create-a-load-balancer-pool

File: internal/cmp/client.go [hunk 3]
```diff
@@ -7,33 +7,39 @@ import apiClient "github.com/HewlettPackard/hpegl-vmaas-cmp-go-sdk/pkg/client"
...
-	TransportZone           DataSource
+	Instance                  Resource
+	InstanceClone             Resource
+	Router                    Resource
+	ResNetwork                Resource
+	RouterNat                 Resource
+	RouterFirewallRuleGroup   Resource
+	RouterRoute               Resource
+	RouterBgpNeighbor         Resource
+	LoadBalancer              Resource
```
Comment (manjunath-batakurki) [inline internal/cmp/client.go:18]: add a data source for load balancer

File: internal/cmp/lb_pools.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	if err := tftags.Get(d, &lbPoolResp); err != nil {
+		return err
+	}
+
+	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	getlbPoolResp, err := lb.lbClient.GetSpecificLBPool(ctx, lbDetails.GetNetworkLoadBalancerResp[0].ID, lbPoolResp.ID)
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_pools.go:37]: lbDetails.GetNetworkLoadBalancerResp[0].ID, we cant assume this, please get this from user

File: internal/cmp/lb_pools.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	if err := tftags.Get(d, &createReq.CreateLBPoolReq); err != nil {
+		return err
+	}
+
+	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	createReq.CreateLBPoolReq.PoolConfig.SnatTranslationType = "LBSnatAutoMap"
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_pools.go:79]: are we not getting this from user ?

File: internal/cmp/lb_profiles.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	if err := tftags.Get(d, &lbProfileResp); err != nil {
+		return err
+	}
+
+	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	getlbProfileResp, err := lb.lbClient.GetSpecificLBProfile(ctx, lbDetails.GetNetworkLoadBalancerResp[0].ID, lbProfileResp.ID)
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_profiles.go:37]: ditto

File: internal/cmp/lb_profiles.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	if err := tftags.Get(d, &createReq.CreateLBProfileReq); err != nil {
+		return err
+	}
+
+	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	createReq.CreateLBProfileReq.ProfileConfig.ProfileType = "application-profile"
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_profiles.go:84]: why are we setting this values ? are these default values ? Can we not set in the resource definition ?

File: internal/cmp/lb_virtual_servers.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+		return err
+	}
+
+	profileData, err := lb.lbClient.GetLBProfiles(ctx, lbDetails.GetNetworkLoadBalancerResp[0].ID)
+	if err != nil {
+		return err
+	}
+
+	for i, profile := range profileData.GetLBProfilesResp {
+		if profile.LBProfileConfig.ProfileType == "application-profile" &&
```
Comment (manjunath-batakurki) [inline internal/cmp/lb_virtual_servers.go:89]: make string as a const variable

File: internal/cmp/loadbalancer.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	types := typeResp.(models.GetLoadBalancerTypes)
+
+	for i, n := range types.LoadBalancerTypes {
+		if n.Name == nsxt {
+			createReq.NetworkLoadBalancer.Type = types.LoadBalancerTypes[i].Name
+			createReq.NetworkLoadBalancer.NetworkServerID = 1
+			break
+		}
+	}
+	createReq.NetworkLoadBalancer.Config.Loglevel = "INFO"
```
Comment (manjunath-batakurki) [inline internal/cmp/loadbalancer.go:85]: why are we setting this ? are we not taking it from user ?

File: internal/resources/data_source_lb_monitors.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func LBMonitorData() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Computed:    true,
```
Comment (manjunath-batakurki) [inline internal/resources/data_source_lb_monitors.go:20]: name is also Computed field ? it is not given by user ?

File: internal/resources/resource_lb_virtual_servers.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"vip_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "vip_name of Network loadbalancer virtual server name",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (manjunath-batakurki) [inline internal/resources/resource_lb_virtual_servers.go:25]: is required and forceNew needed here ?

File: internal/resources/resource_lb_virtual_servers.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+				Description: "description of Network loadbalancer virtual server",
+				ForceNew:    true,
+			},
+			"vip_address": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "vip_address of Network loadbalancer virtual server",
+				ForceNew:    true,
```
Comment (manjunath-batakurki) [inline internal/resources/resource_lb_virtual_servers.go:33]: if the user changes this, it will create a new virtual server ?

**Comment by Pushpalatha555** [comment]
@manjunath-batakurki All the comments I can take it in next PR still load balancer not yet finished we have 2more stories to work to complete the load balancer.

File: internal/resources/resource_lb_virtual_servers.go (hunk 11 above)
Comment (Pushpalatha555) [reply internal/resources/resource_lb_virtual_servers.go:25]: done

File: internal/resources/resource_lb_virtual_servers.go (hunk 12 above)
Comment (Pushpalatha555) [reply internal/resources/resource_lb_virtual_servers.go:33]: done

File: internal/resources/data_source_lb_monitors.go (hunk 10 above)
Comment (Pushpalatha555) [reply internal/resources/data_source_lb_monitors.go:20]: done

File: internal/cmp/loadbalancer.go (hunk 9 above)
Comment (Pushpalatha555) [reply internal/cmp/loadbalancer.go:85]: updated

File: internal/cmp/loadbalancer.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func (lb *loadBalancer) Create(ctx context.Context, d *utils.Data, meta interface{}) error {
+	setMeta(meta, lb.lbClient.Client)
+	createReq := models.CreateLoadBalancerRequest{
+		NetworkLoadBalancer: models.CreateNetworkLoadBalancerRequest{
+			Name:        d.GetString("name"),
+			Description: d.GetString("description"),
+			Enabled:     d.GetBool("enabled"),
+			Visibility:  d.GetString("visibility"),
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/cmp/loadbalancer.go:50]: removed

File: internal/cmp/lb_virtual_servers.go (hunk 8 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_virtual_servers.go:89]: done

File: internal/cmp/lb_profiles.go (hunk 7 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_profiles.go:84]: done

File: internal/cmp/lb_profiles.go (hunk 6 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_profiles.go:37]: done

File: internal/cmp/lb_pools.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func (lb *loadBalancerPool) Delete(ctx context.Context, d *utils.Data, meta interface{}) error {
+	lbPoolID := d.GetID()
+	lbDetails, err := lb.lbClient.GetLoadBalancers(ctx)
+	if err != nil {
+		return err
+	}
+
+	_, err = lb.lbClient.DeleteLBPool(ctx, lbDetails.GetNetworkLoadBalancerResp[0].ID, lbPoolID)
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/cmp/lb_pools.go:116]: done

File: internal/cmp/lb_pools.go (hunk 5 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_pools.go:79]: updated

File: internal/cmp/lb_pools.go (hunk 4 above)
Comment (Pushpalatha555) [reply internal/cmp/lb_pools.go:37]: done

File: internal/cmp/client.go (hunk 3 above)
Comment (Pushpalatha555) [reply internal/cmp/client.go:18]: done

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb_pool.tf (hunk 2 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb_pool.tf:3]: done

File: examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf (hunk 1 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_load_balancer/nsx_t_lb.tf:7]: removed

File: internal/resources/resource_lb_monitors.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+			"description": {
+				Type:        schema.TypeString,
+				Description: "Creating the Network Load balancer Monitor.",
+				Optional:    true,
+			},
+			"send_type": {
+				Type:        schema.TypeString,
+				Description: "send type method like GET,POST",
+				Optional:    true,
+			},
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/resources/resource_lb_monitors.go:32]: done

**Pull request merged**